[sbt 1.0] Use Scala 2.12.1

eed3si9n opened the pull request:
This is a PR on top of https://github.com/sbt/sbt/pull/2907.

dwijnand commented:
```

[warn] /home/travis/build/sbt/sbt/main/src/test/scala/Delegates.scala:57: local val in method globalCombinations is never used
[warn]       val value = axis(s)
[warn]           ^
[warn] /home/travis/build/sbt/sbt/main/src/test/scala/PluginsTest.scala:34: possible missing interpolator: detected interpolated identifier `$S`
[warn]       deducePlugin(S, log) must throwAn[AutoPluginException](message = """Contradiction in enabled plugins:
[warn]                                                                        ^
[warn] /home/travis/build/sbt/sbt/main/src/test/scala/PluginsTest.scala:40: possible missing interpolator: detected interpolated identifier `$T`
[warn]       deducePlugin(T && U, log) must throwAn[AutoPluginException](message = """Contradiction in enabled plugins:
[warn]                                                                             ^
[error] /home/travis/build/sbt/sbt/main/src/test/scala/sbt/internal/TestBuild.scala:249: not enough arguments for method apply: (p: org.scalacheck.Gen.Parameters, seed: org.scalacheck.rng.Seed)Option[T] in class Gen.
[error] Unspecified value parameter seed.
[error]     wrap(gs map { g => g(prms) getOrElse sys.error("failed generator") })
[error]                         ^
[warn] three warnings found
[error] one error found
[error] (mainProj/test:compileIncremental) Compilation failed
[error] (actionsProj/test:test) sbt.TestsFailedException: Tests unsuccessful
```

eed3si9n commented:
Minor Scalacheck touchups were required - 6703949

eed3si9n committed:
Use Scala 2.12.1

Bump up the Scala version used by sbt to 2.12.1.
Most changes are migration of the integration tests.

eed3si9n commented:
All tests are green, so I squashed the commits.

dwijnand commented on main/src/test/scala/ParseKey.scala:
```diff
@@ -137,5 +175,8 @@ object ParseKey extends Properties("Key parser test") {
-    pick((load * from.size).toInt, from)
+    pick((load * from.size).toInt match {
+      case 0 => 1
+      case x => x
```
Could you explain what's happening here, and clean up this file please?

eed3si9n replied:
`pickN` is a function that randomly picks `load` % items from the `from` sequence, but the rest of the tests expect at least one item, so I changed it to return 1 in case of 0. In grand scheme of things, it's a random build generator for tests, and I don't think it compromises the coverage of the test.

eed3si9n replied:
This happened because I had to bump ScalaCheck 1.11.4 to 1.13.4.

intracer replied:
As an alternative - `Math.max(1, x)` ?

eed3si9n replied:
Yea, that sounds better.

eed3si9n commented:
Addressed the review comments, so merging this.

dwijnand commented on main/src/test/scala/ParseKey.scala:
```diff
@@ -136,6 +136,8 @@ object ParseKey extends Properties("Key parser test") {
+  // The rest of the tests expect at least one item, so I changed it to return 1 in case of 0.
   def pickN[T](load: Double, from: Seq[T]): Gen[Seq[T]] =
-    pick((load * from.size).toInt, from)
+    pick(Math.max((load * from.size).toInt, 1), from)
```
Could you explain why this is suddenly necessary with Scala 2.12.1?

eed3si9n replied:
It's not Scala 2.12.1 per se, but the upgrade of ScalaCheck that seems to have changed something.

dwijnand replied:
Gotcha. Thanks.

dwijnand commented on main/src/test/scala/sbt/internal/TestBuild.scala:
```diff
@@ -170,6 +178,15 @@ object TestBuild {
+        } catch {
+          case e: Throwable =>
+            e.printStackTrace
+            throw e
```
Also, are these also necessary?

eed3si9n replied:
Probably not.
